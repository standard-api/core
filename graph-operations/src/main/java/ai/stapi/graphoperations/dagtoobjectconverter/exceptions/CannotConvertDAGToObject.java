package ai.stapi.graphoperations.dagtoobjectconverter.exceptions;

import ai.stapi.graph.traversableGraphElements.TraversableNode;

public class CannotConvertDAGToObject extends RuntimeException {

    private CannotConvertDAGToObject(String message) {
        super("Cannot convert DAG to Object, because " + message);
    }
    
    public static CannotConvertDAGToObject becauseItContainsCycle(TraversableNode twiceVisitedNode) {
        return new CannotConvertDAGToObject(
            String.format(
                "it contains cycle. Twice visited node: '%s/%s'",
                twiceVisitedNode.getType(),
                twiceVisitedNode.getId()
            )
        );
    }
}
