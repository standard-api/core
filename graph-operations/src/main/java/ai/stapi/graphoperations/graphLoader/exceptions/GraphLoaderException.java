package ai.stapi.graphoperations.graphLoader.exceptions;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.utils.Stringifier;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;

public class GraphLoaderException extends RuntimeException {

  public GraphLoaderException(String becauseMessage) {
    super("Cannot load graph, " + becauseMessage);
  }

  private GraphLoaderException(String becauseMessage, Throwable cause) {
    super("Cannot load graph, " + becauseMessage, cause);
  }

  public static GraphLoaderException becauseThereIsNoGraphElementWithProvidedUuid(
      UniqueIdentifier uuid,
      GraphDescription graphDescription
  ) {
    return new GraphLoaderException(
        "because specified graph element does not exist in graph." +
            "\nUUID: " + uuid +
            "\nGraphDescription: " + graphDescription.getClass().getSimpleName() +
            "\n" + Stringifier.convertToString(graphDescription)
    );
  }

  public static GraphLoaderException becauseThereIsMoreThanOneNodeDescriptionsOnEdgeDescription(
      UniqueIdentifier id, String type) {
    return new GraphLoaderException(
        "because there is more than one node descriptions on edge description." +
            "\nUnique Identifier: " + id +
            "\nEdge type: " + type
    );
  }

  public static GraphLoaderException becauseEdgeDescriptionHadMoreChildNodeDescriptions(
      AbstractEdgeDescription edgeDescription,
      List<String> nodeTypes
  ) {
    var param = (EdgeDescriptionParameters) edgeDescription.getParameters();
    return new GraphLoaderException(
        "because edge description had more child node descriptions." +
            "\nEdge Description ClassName: " + edgeDescription.getClass().getSimpleName() +
            "\nEdge type: " + param.getEdgeType() +
            "\nNode Types: " + String.join(", ", nodeTypes)
    );
  }

  public static GraphLoaderException becauseProvidedOGMDidNotStartWithObjectOGM(
      ObjectGraphMapping objectGraphMapping) {
    return new GraphLoaderException(
        "because provided OGM did not start with object OGM." +
            "\nOGM Class: " + objectGraphMapping.getClass().getSimpleName()
    );
  }
}
