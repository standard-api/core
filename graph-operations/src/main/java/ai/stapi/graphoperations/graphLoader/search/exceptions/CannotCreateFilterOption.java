package ai.stapi.graphoperations.graphLoader.search.exceptions;

import ai.stapi.graphoperations.graphLoader.search.filterOption.FilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.FilterOptionParameters;

public class CannotCreateFilterOption extends RuntimeException {

  private CannotCreateFilterOption(String becauseMessage) {
    super("Cannot create FilterOption, " + becauseMessage);
  }

  public static CannotCreateFilterOption becauseInvalidLeafFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid leaf filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidListFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid list filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidLogicalFilterStrategyProvided(
      String filterStrategy) {
    return new CannotCreateFilterOption(
        "because invalid logical filter strategy provided. " +
            "Strategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidFilterOptionParameters(
      String filterStrategy,
      FilterOptionParameters filterOptionParameters
  ) {
    return new CannotCreateFilterOption(
        "because invalid filter option parameters provided. " +
            "\nParameters: " + filterOptionParameters.getClass().getSimpleName() +
            "\nStrategy: " + filterStrategy
    );
  }

  public static CannotCreateFilterOption becauseInvalidFilterOptionFoundWhenMakingCopy(
      FilterOption<?> originalFilter) {
    return new CannotCreateFilterOption(
        "because invalid filter option found when making copy. " +
            "\nFilter option: " + originalFilter.getClass().getSimpleName()
    );
  }
}
