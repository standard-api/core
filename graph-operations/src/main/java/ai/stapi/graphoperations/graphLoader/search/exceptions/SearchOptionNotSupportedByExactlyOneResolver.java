package ai.stapi.graphoperations.graphLoader.search.exceptions;

import ai.stapi.graphoperations.graphLoader.search.SearchOption;

public class SearchOptionNotSupportedByExactlyOneResolver extends RuntimeException {

  public SearchOptionNotSupportedByExactlyOneResolver(int size, SearchOption<?> searchOption) {
    super(
        String.format(
            "Search Option Of type: \"%s\" is supported by %d resolvers. But it should be exactly one.",
            searchOption.getClass().getCanonicalName(),
            size
        )
    );
  }
}
