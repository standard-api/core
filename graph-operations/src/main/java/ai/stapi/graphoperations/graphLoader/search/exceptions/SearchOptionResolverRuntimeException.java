package ai.stapi.graphoperations.graphLoader.search.exceptions;

public class SearchOptionResolverRuntimeException extends RuntimeException {

  private SearchOptionResolverRuntimeException(String becauseMessage) {
    super("Search option resolver threw exception, " + becauseMessage);
  }

  public static SearchOptionResolverRuntimeException becauseSearchQueryParamatersCanOnlyHaveOnePagingOptions() {
    return new SearchOptionResolverRuntimeException(
        "because Search Query Paramaters Can Only Have One Paging Options"
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideSortOptionMustBeSinglePath() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside Sort Option must be single path."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideSortOptionMustEndWithValueDescription() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside Sort Option must end with value description."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideOneValueFilterOptionMustBeSinglePath() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside OneValueFilterOption must be single path."
    );
  }

  public static SearchOptionResolverRuntimeException becauseGraphDescriptionInsideOneValueFilterOptionMustEndWithValueDescription() {
    return new SearchOptionResolverRuntimeException(
        "because GraphDescription inside OneValueFilterOption must end with value description."
    );
  }

}
