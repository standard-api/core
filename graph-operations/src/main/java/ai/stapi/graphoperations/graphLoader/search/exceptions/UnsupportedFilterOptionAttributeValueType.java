package ai.stapi.graphoperations.graphLoader.search.exceptions;

public class UnsupportedFilterOptionAttributeValueType extends RuntimeException {

  public UnsupportedFilterOptionAttributeValueType(
      Class<?> attributeValueType,
      Class<?> optionClass
  ) {
    super(
        String.format(
            "Attribute value of type: \"%s\" cannot be used in: \"%s\"",
            attributeValueType.getCanonicalName(),
            optionClass.getCanonicalName()
        )
    );
  }

  public UnsupportedFilterOptionAttributeValueType(
      Class<?> optionClass
  ) {
    super(
        String.format(
            "Null attribute value cannot be used in: \"%s\"",
            optionClass.getCanonicalName()
        )
    );
  }
}
