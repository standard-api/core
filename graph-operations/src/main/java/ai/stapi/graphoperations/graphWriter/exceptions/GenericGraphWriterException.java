package ai.stapi.graphoperations.graphWriter.exceptions;

import ai.stapi.graphoperations.declaration.Declaration;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.removal.RemovalGraphDescription;
import ai.stapi.graphoperations.graphWriter.GenericGraphWriter;
import ai.stapi.graphoperations.graphWriter.SpecificGraphWriter;
import ai.stapi.utils.LineFormatter;
import java.util.List;
import java.util.stream.Collectors;

public class GenericGraphWriterException extends RuntimeException {

  public GenericGraphWriterException(String format) {
    super(format);
  }

  public static GenericGraphWriterException becauseNoSupportingSpecificResolverForGivenDeclaration(
      Declaration declaration) {
    return new GenericGraphWriterException(
        String.format(
            "There is no supporting %s for declaration '%s'.",
            SpecificGraphWriter.class.getSimpleName(),
            declaration.getClass().getSimpleName()
        )
    );
  }

  public static GenericGraphWriterException becauseMoreThanOneSpecificResolverForGivenDeclaration(
      Declaration declaration,
      List<SpecificGraphWriter> supportingSpecificGraphWriters
  ) {
    return new GenericGraphWriterException(
        String.format(
            "There is more than one supporting %s for declaration '%s'."
                + LineFormatter.createNewLine() +
                "Supporting: " + LineFormatter.createNewLine() +
                "'%s'",
            SpecificGraphWriter.class.getSimpleName(),
            declaration.getClass().getSimpleName(),
            supportingSpecificGraphWriters.stream()
                .map(supportingSpecificGraphWriter -> supportingSpecificGraphWriter.getClass()
                    .getSimpleName())
                .collect(Collectors.joining("'" + LineFormatter.createNewLine() + "'"))
        )
    );
  }

  public static GenericGraphWriterException becauseFirstDescriptionHasToBeNode(
      PositiveGraphDescription graphDescription) {
    return new GenericGraphWriterException(
        "First graph description must be of type '"
            + NodeDescription.class.getSimpleName()
            + "' but was of type '"
            + graphDescription.getClass().getSimpleName()
            + "'."
    );
  }

  public static GenericGraphWriterException becauseGraphDescriptionContainsRemovalGraphDescription(
      GraphDescription description) {
    return new GenericGraphWriterException(
        "Provided "
            + GraphDescription.class.getSimpleName()
            + " contains " + RemovalGraphDescription.class.getSimpleName()
            + " of type '" + description.getClass().getSimpleName()
            + "' and that is not allowed in " + GenericGraphWriter.class.getSimpleName()
            + "."
    );
  }
}
