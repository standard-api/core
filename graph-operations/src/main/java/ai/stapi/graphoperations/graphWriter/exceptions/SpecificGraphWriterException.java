package ai.stapi.graphoperations.graphWriter.exceptions;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.UuidIdentityDescription;
import ai.stapi.graphoperations.graphWriter.specific.AttributeGraphWriter;

public class SpecificGraphWriterException extends RuntimeException {

  private SpecificGraphWriterException(String message) {
    super(message);
  }

  public static SpecificGraphWriterException becauseGivenEdgeDirectionIsNotSupported(
      GraphDescription graphDescription) {
    return new SpecificGraphWriterException(
        "Given edge direction '"
            + graphDescription.getClass().getSimpleName()
            + "' is not supported."
    );
  }

  public static GenericGraphWriterException becauseEdgeCanNotBeFirstElementInGraph() {
    return new GenericGraphWriterException("Graph can not start with edge.");
  }

  public static SpecificGraphWriterException becauseAttributeDescriptionDoesNotContainValue(
      GraphDescription graphDescription) {
    return new SpecificGraphWriterException(
        "Provided attribute description '"
            + graphDescription.getClass().getSimpleName()
            + "' does not contain value."
    );
  }

  public static SpecificGraphWriterException becauseAttributeTypeIsNotSupported(
      GraphDescription graphDescription) {
    return new SpecificGraphWriterException(
        "Provided attribute description '"
            + graphDescription.getClass().getSimpleName()
            + "' is not recognized by '" + AttributeGraphWriter.class.getSimpleName()
            + "'. Add it to 'getAttributeValue' method."
    );
  }

  public static SpecificGraphWriterException becauseProvidedConstantIsNotUuid(Object value) {
    return new SpecificGraphWriterException(
        "Provided constant is not of type UUID. Provided constant: '" + value + "'."
    );
  }

  public static SpecificGraphWriterException becauseProvidedAttributeValueDoesNotContainAnyConstantDescription(
      GraphDescription graphDescription) {
    return new SpecificGraphWriterException(
        "'" + graphDescription.getClass().getSimpleName()
            + "' does not contain any constant description among it's children."
    );
  }

  public static SpecificGraphWriterException becauseRemovalDescriptionDoesNotHaveUuidDescriptionAsChild(
      GraphDescription graphDescription
  ) {
    return new SpecificGraphWriterException(
        "Removal description '"
            + graphDescription.getClass().getSimpleName()
            + "' does not contain any '"
            + UuidIdentityDescription.class.getSimpleName()
            + "' as child ,and that is not allowed."
    );
  }
}
