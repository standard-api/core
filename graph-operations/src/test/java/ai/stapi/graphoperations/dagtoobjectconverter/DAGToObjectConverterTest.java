package ai.stapi.graphoperations.dagtoobjectconverter;

import ai.stapi.graph.Graph;
import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.BooleanAttributeValue;
import ai.stapi.graph.attribute.attributeValue.IntegerAttributeValue;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graphoperations.dagtoobjectconverter.exceptions.CannotConvertDAGToObject;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

class DAGToObjectConverterTest extends SchemaIntegrationTestCase {
    
    @Autowired
    private DAGToObjectConverter dagToObjectConverter;

    @Test
    void itWillThrowExceptionWhenCyclicGraphProvided() {
        var a = new Node(new UniqueIdentifier("id-1"), "A");
        var b = new Node(new UniqueIdentifier("id-2"),"B");
        var givenCyclicGraph = new Graph(a, b, new Edge(a, "1", b), new Edge(b, "2", a));
        var givenNode = givenCyclicGraph.traversable().loadNode(
            a.getId(),
            a.getType()
        );
        Executable throwable = () -> this.dagToObjectConverter.convert(givenNode);
        this.thenExceptionMessageApproved(CannotConvertDAGToObject.class, throwable);
    }

    @Test
    void itWillThrowExceptionWhenBiggerCyclicGraphProvided() {
        var a = new Node(new UniqueIdentifier("id-1"),"A");
        var b = new Node(new UniqueIdentifier("id-2"),"B");
        var c = new Node(new UniqueIdentifier("id-3"),"C");
        var d = new Node(new UniqueIdentifier("id-4"),"D");
        var e = new Node(new UniqueIdentifier("id-5"),"E");
        var e1 = new Edge(a, "1", b);
        var e2 = new Edge(b, "2", c);
        var e3 = new Edge(c, "3", d);
        var e4 = new Edge(d, "4", e);
        var e5 = new Edge(e, "5", a);
        var givenCyclicGraph = new Graph(a, b, c, d, e, e1, e2, e3, e4, e5);
        var givenNode = givenCyclicGraph.traversable().loadNode(c.getId(), c.getType());
        Executable throwable = () -> this.dagToObjectConverter.convert(givenNode);
        this.thenExceptionMessageApproved(CannotConvertDAGToObject.class, throwable);
    }
    
    @Test
    void itCanConvertSimpleDAG() {
        var house = new Node(
            new UniqueIdentifier("id-1"),
            "House",
            new LeafAttribute<>("title", new StringAttributeValue("mohykan")),
            new LeafAttribute<>("height", new IntegerAttributeValue(500))
        );
        var door = new Node(
            new UniqueIdentifier("id-2"),
            "Door",
            new LeafAttribute<>("closable", new BooleanAttributeValue(false)),
            new LeafAttribute<>("width", new IntegerAttributeValue(20))
        );
        var door2 = new Node(
            new UniqueIdentifier("id-3"),
            "Door",
            new LeafAttribute<>("closable", new BooleanAttributeValue(true)),
            new LeafAttribute<>("width", new IntegerAttributeValue(15))
        );
        var handle = new Node(
            new UniqueIdentifier("id-4"),
            "Handle",
            new LeafAttribute<>("color", new StringAttributeValue("red"))
        );
        var edge1 = new Edge(house, "hasDoor", door);
        var edge2 = new Edge(house, "hasDoor", door2);
        var edge3 = new Edge(door, "hasHandle", handle);
        var givenGraph = new Graph(
            house,
            door,
            door2,
            handle,
            edge1,
            edge2,
            edge3
        );
        var givenNode = givenGraph.traversable().loadNode(house.getId(), house.getType());
        var actualObject = this.dagToObjectConverter.convert(givenNode);
        this.thenObjectApproved(actualObject);
    }
}