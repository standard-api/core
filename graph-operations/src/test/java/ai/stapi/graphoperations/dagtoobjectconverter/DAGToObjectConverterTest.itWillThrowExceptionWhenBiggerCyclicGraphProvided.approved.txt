Cannot convert DAG to Object, because it contains cycle. Twice visited node: 'C/id-3'